service/locker/distributed.go

```diff
@@ -0,0 +1,106 @@
+package locker
+
+import (
+	"context"
+	"time"
+
+	"github.com/giantswarm/kubelock"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+
+	"github.com/giantswarm/aws-operator/pkg/project"
+)
+
+const (
+	lockName          = "ipam"
+	lockNamespaceName = "giantswarm"
+)
+
+var (
+	lockOwner = project.Name() + "@" + project.Version()
+	lockTTL   = 30 * time.Second
+)
+
+type DistributedLockerConfig struct {
+	Logger     micrologger.Logger
+	RestConfig *rest.Config
+}
+
+type DistributedLocker struct {
+	logger micrologger.Logger
+
+	kubelock kubelock.Interface
+}
+
+func NewDistributedLocker(config DistributedLockerConfig) (*DistributedLocker, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.RestConfig == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.RestConfig must not be empty", config)
+	}
+
+	var err error
+	var gvr schema.GroupVersionResource
+	{
+		gvr = schema.GroupVersionResource{
+			Group:    "",
+			Version:  "v1",
+			Resource: "namespaces",
+		}
+	}
+
+	var dynClient dynamic.Interface
+	{
+		dynClient, err = dynamic.NewForConfig(config.RestConfig)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+	}
+
+	var k kubelock.Interface
+	{
+		c := kubelock.Config{
+			DynClient: dynClient,
+			GVR:       gvr,
+		}
+		k, err = kubelock.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	d := &DistributedLocker{
+		logger: config.Logger,
+
+		kubelock: k,
+	}
+
+	return d, nil
+}
+
+func (d DistributedLocker) Lock(ctx context.Context) error {
+	err := d.kubelock.Lock(lockName).Acquire(ctx, lockNamespaceName, kubelock.AcquireOptions{
+		Owner: lockOwner,
+		TTL:   lockTTL,
```

Comment:
How do we actually ensure that the action for which we want to lock is not timing out while we are operating?

Reply:
Do you mean how we ensure that lock doesn't timeout before the guarded operation finishes?

Reply:
Exactly. Like heartbeating or something.

Reply:
But I could imagine that our operation takes like 5 seconds and the TTL is 30 minutes. Then this should be safe enough for us for now. Just curious how we thought to address this eventually. /cc @kopiczko

Reply:
Our operation takes under 1s and we have a 30s lock here, we are more than safe.